### PR TITLE
[HINTS]Add implementation for hints on function is_zero (cairo_secp/field.cairo )

### PR DESCRIPTION
**Pull request by pefontana**
# [HINTS]Add implementation for hints on function is_zero (cairo_secp/field.cairo )

## Description

[cairo_secp/field.cairo](https://github.com/starkware-libs/cairo-lang/blob/167b28bcd940fd25ea3816204fa882a0b0a49603/src/starkware/cairo/common/cairo_secp/field.cairo#L97)

* Add SECP_P static item to SecpUtils.
* Update MathUtils::div_mod fn `div_mod(a, b, prime) ->  div_mod(a, b, &prime)`

* Hints implementation:

```
is_zero_pack:
%{
    from starkware.cairo.common.cairo_secp.secp_utils import SECP_P, pack

    x = pack(ids.x, PRIME) % SECP_P
%}
```

```
is_zero_nondet:
if nondet %{ x == 0 %} != 0:
```

```
is_zero_assign_scope_variables:
%{
    from starkware.cairo.common.cairo_secp.secp_utils import SECP_P
    from starkware.python.math_utils import div_mod

    value = x_inv = div_mod(1, x, SECP_P)
%}
```

Description of the pull request changes and motivation.

## Checklist
- [ ] Linked to Github Issue
- [x] Unit tests added
- [x] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

File: src/vm/hints/secp/field_utils.rs [hunk 1]
```diff
@@ -78,30 +74,117 @@ pub fn reduce(
...
+    {
+        Some(PyValueType::BigInt(x)) => x,
+        _ => {
+            return Err(VirtualMachineError::VariableNotInScopeError(String::from(
+                "x",
+            )))
+        }
+    };
+
+    let value = if x.is_zero() { bigint!(1) } else { bigint!(0) };
```
Comment (Oppen) [inline src/vm/hints/secp/field_utils.rs:135]:
```suggestion
    let value = bigint!(x.is_zero() as usize);
```

File: src/vm/hints/secp/field_utils.rs [hunk 2]
```diff
@@ -553,4 +636,360 @@ mod tests {
             ))
         );
     }
+
+    #[test]
+    fn run_is_zero_pack_ok() {
+        let hint_code = "from starkware.cairo.common.cairo_secp.secp_utils import SECP_P, pack\n\nx = pack(ids.x, PRIME) % SECP_P".as_bytes();
+        let mut vm = VirtualMachine::new(
+            BigInt::new(Sign::Plus, vec![1, 0, 0, 0, 0, 0, 17, 134217728]),
```
Comment (Oppen) [inline src/vm/hints/secp/field_utils.rs:644]: We can replace this with a module level `bigint_str` for legibility and reusability.

File: src/vm/hints/secp/secp_utils.rs [hunk 3]
```diff
@@ -8,6 +8,14 @@ use num_traits::{Signed, Zero};
 lazy_static! {
     pub static ref BASE_86: BigInt = bigint!(1) << 86_usize;
     pub static ref BASE_86_MAX: BigInt = &*BASE_86 - bigint!(1);
+    pub static ref SECP_P: BigInt = (bigint!(2) << (255))
```
Comment (Oppen) [inline src/vm/hints/secp/secp_utils.rs:11]:
```suggestion
    pub static ref SECP_P: BigInt = (bigint!(1) << (256))
```

**Review by Oppen — CHANGES_REQUESTED** [review]

File: src/vm/hints/secp/field_utils.rs (hunk 2 above)
Comment (pefontana) [reply src/vm/hints/secp/field_utils.rs]: True,
I created a `vm_prime!()` macro, tell me what you think
https://github.com/lambdaclass/cleopatra_cairo/pull/332/commits/216a6e148595ea1545f5c376b6f1fdc564865c38

File: src/vm/hints/secp/secp_utils.rs (hunk 3 above)
Comment (pefontana) [reply src/vm/hints/secp/secp_utils.rs]: Done
https://github.com/lambdaclass/cleopatra_cairo/pull/332/commits/4e34be230d1933df9a0b5358e8daab847d310807

File: src/vm/hints/secp/field_utils.rs (hunk 1 above)
Comment (pefontana) [reply src/vm/hints/secp/field_utils.rs]: Done
https://github.com/lambdaclass/cleopatra_cairo/pull/332/commits/4e34be230d1933df9a0b5358e8daab847d310807

File: src/utils.rs [hunk 4]
```diff
@@ -16,6 +16,17 @@ macro_rules! bigint_str {
...
 
+#[macro_export]
+macro_rules! vm_prime {
+    () => {
+        BigInt::parse_bytes(
+            b"3618502788666131213697322783095070105623107215331596699973092056135872020481",
+            10,
+        )
+        .unwrap()
+    };
```
Comment (Oppen) [inline src/utils.rs:27]: A `lazy_static!` constant makes more sense for this.

**Review by Oppen — CHANGES_REQUESTED** [review]

File: src/utils.rs (hunk 4 above)
Comment (pefontana) [reply src/utils.rs]: Yes,
The only thing I don't like about the `lazy_static!` is that we have to use the `VM_PRIME.clone()` in all the tests
https://github.com/lambdaclass/cleopatra_cairo/pull/332/commits/2600683f6541c87f3aea44b975918b66b09123ac

File: src/vm/hints/secp/field_utils.rs [hunk 5]
```diff
@@ -78,36 +74,122 @@ pub fn reduce(
...
+        .ok_or(VirtualMachineError::ScopeError)?
+        .get("x")
+    {
+        Some(PyValueType::BigInt(x)) => x,
+        _ => {
+            return Err(VirtualMachineError::VariableNotInScopeError(String::from(
+                "x",
+            )))
+        }
+    };
```
Comment (Oppen) [inline src/vm/hints/secp/field_utils.rs:133]: I think one of the helpers @fmoletta added can make this more brief.

**Review by Oppen — CHANGES_REQUESTED** [review]

File: src/vm/hints/secp/field_utils.rs (hunk 5 above)
Comment (pefontana) [reply src/vm/hints/secp/field_utils.rs]: Nice, this will simplify things even more. 
Done

**Review by Oppen — APPROVED** [review]

**Pull request merged**